torch/distributed/fsdp/fully_sharded_data_parallel.py

```diff
@@ -2003,6 +2014,300 @@ def clip_grad_norm_(
                 assert p.grad is not None
                 p.grad.detach().mul_(clip_coef.to(p.grad.device))
 
+    @staticmethod
+    def full_optim_state_dict(
+        model: torch.nn.Module,
+        optim: torch.optim.Optimizer,
+        optim_input: Optional[Union[
+            List[Dict[str, Any]], Iterable[torch.nn.Parameter],
+        ]] = None,
+    ) -> Dict[str, Any]:
+        """
+        Consolidates the full optimizer state on rank 0 and returns it
+        as a :class:`dict` following the convention of
+        :meth:`torch.optim.Optimizer.state_dict`, i.e. with keys ``"state"``
+        and ``"param_groups"``. The flattened parameters in ``FSDP`` modules
+        contained in ``model`` are mapped back to their unflattened parameters.
+
+        .. warning:: This needs to be called on all ranks since synchronization
+            primitives are used.
+
+        .. warning:: The parameter IDs used in the state dict depend on the
+        ``FSDP`` wrapping scheme used in the passed-in ``model``. This means
+        that they may not match those of a purely non-wrapped version of the
+        model and that the returned state dict cannot necessarily be loaded
+        for a version of the model with a different ``FSDP`` wrapping scheme.
```

Comment:
(can be done in follow up PRs) what's the suggested remedy? Or is there a way to map FSDP param ids back to the original models corresponding param ids?

Reply:
One requirement (can be addressed later) is that users might train the model using FSDP and then load the saved model for inference. The inference could be single device or multi-device pipeline. To support this, we might need an additional helper functions, or let all distributed training/inference features share the same saved format.

Reply:
In my understanding, for inference, the optimizer state is no longer relevant and does not need to be loaded.

The challenge has been how to match optimizer state tensors to their corresponding unflattened parameters. This is good news because model state dict does not need to address this challenge since it works with the unflattened parameters directly. (There is no matching process.) I have not thought about it in depth, but at the least, model state checkpointing should not suffer from the same blocker for changing the FSDP wrapping (which would be required for inference).

Reply:
To follow-up on this, it looks like using parameter names has resolved this issue.